src/sage/data_structures/stream.py

```diff
@@ -999,6 +1024,38 @@ def __init__(self, function, is_sparse, approximate_order, true_order=False):
         super().__init__(is_sparse, true_order)
         self._approximate_order = approximate_order
 
+    def input_streams(self):
+        r"""
+        Return the list of streams which are used to compute the
+        coefficients of ``self``, as provided.
+
+        EXAMPLES:
+
+        Only streams that appear in the closure are detected::
+
+            sage: from sage.data_structures.stream import Stream_function, Stream_exact
+            sage: f = Stream_exact([1,3,5], constant=7)
+            sage: g = Stream_function(lambda n: f[n]^2, False, 0)
+            sage: g.input_streams()
+            []
+
+            sage: def fun():
+            ....:     f = Stream_exact([1,3,5], constant=7)
+            ....:     g = Stream_function(lambda n: f[n]^2, False, 0)
+            ....:     return g.input_streams()
+            sage: fun()
+            [<sage.data_structures.stream.Stream_exact object at 0x...>]
+        """
+        closure = self.get_coefficient.__closure__
```

Comment:
This code is so hacky and liable to pick up far more than it actually needs and likely could be quite slow because it needs to check potentially a lot of objects. I think we should consider caching the result. I would still return a `list` for consistency, but with a warning to the user not to mutate it (since it isn't part of the public facing stuff, only a more serious developer will get into this). Although perhaps that is slightly dangerous even within our code...? Perhaps these that need to do some computations be cached and every such method returns a `tuple`?

Reply:
This code is only executed once for each operation when *creating* the power series.  For example,
```
sage: L.<z> = LazyPowerSeriesRing(SR)
sage: G = L.undefined(0)
sage: L.define_implicitly([(G, [ln(2)])], [diff(G) - exp(-G(-z))])
sage: %prun G[:20]
         532675 function calls (516627 primitive calls) in 0.815 seconds

   Ordered by: internal time
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
...
        2    0.000    0.000    0.000    0.000 stream.py:1027(input_streams)
```